python/pfs/pipe2d/generateCommands.py

```diff
@@ -0,0 +1,1493 @@
+    def fromYaml(cls, yamlBlock: Mapping[str, Any]) -> "CommandConfig":
+        """Extract "config: ..." from ``yamlBlock``.
+
+        Parameters
+        ----------
+        yamlBlock : `Mapping[str, Any]`
+            A block of a YAML data structure:
+
+            ``"config"``
+                A config string, or a list of config strings.
+            ``"configfile"``
+                Path to a configuration file
+
+        Returns
+        -------
+        config : `CommandConfig`
+        """
+        configs = yamlBlock.get("config", [])
+        if isinstance(configs, str):
+            configs = [configs]
+        if not isinstance(configs, list):
+            raise RuntimeError(f"'config' must be a string or a list of string: ({configs})")
+
+        configs = [ensureKeyEqValue(str(x)) for x in configs]
+
+        configfile = yamlBlock.get("configfile")
+        if not (configfile is None or isinstance(configfile, str)):
+            raise RuntimeError(f"'configfile' must be a string: '{configfile}'")
+
+        return cls(configs=configs, configfile=configfile)
+
+    def getCommandLine(self) -> List[str]:
+        """Get command line options for this config.
+
+        Returns
+        -------
+        options : `List[str]`
+            Like ["--config", "key=value", "key=value"]
+        """
+        commands = []
+        if self.configfile:
+            commands.append(f"--configfile={self.configfile}")
+        if self.configs:
+            commands += ["--config"] + self.configs
+        return commands
+
+
+@export
+class SourceFilter:
+    """This class represents ``--id`` in a command line.
+
+    If two or more arguments are specified in construction,
+    the result will be their intersection.
+
+    Parameters
+    ----------
+    id : `Iterable[str]`
+        Arguments given to ``--id`` in a command line,
+        like ``["field=BIAS"]``
+    """
+
+    __slots__ = ["id"]
+
+    def __init__(self, id: Iterable[str] = []):
+        self.id = list(id)
+
+    @classmethod
+    def fromYaml(cls, yamlBlock: Mapping[str, Any], *, key: str = "id") -> "SourceFilter":
+        """Extract "id:..." from ``yamlBlock``.
+
+        Parameters
+        ----------
+        yamlBlock : `Mapping[str, Any]`
+            A block of a YAML data structure:
+
+            ``"id"``
+                A string like ``"field=BIAS"`` or a list of such strings.
+
+        key : `str`
+            If you give this argument, ``key`` will be used
+            instead of ``"id"`` for the key of the extracted field.
+
+        Returns
+        -------
+        sourceFilter : `SourceFilter`
+        """
+        id = yamlBlock.get(key, [])
+        if isinstance(id, str):
+            id = [id]
+        if not isinstance(id, list):
+            raise RuntimeError(f"'{key}' must be a string or a list of string: '{id}'")
+
+        return cls(ensureKeyEqValue(str(x)) for x in id)
+
+    def getCommandLine(self, *, key: str = "id") -> List[str]:
```

Comment:
Pretty sure spaces around the equals sign are forbidden in the style guide, but not sure how that gets altered when there's types specified. Please make sure that this file passes flake8 with the usual configuration (which you may have to look in drp_stella for).

Reply:
I don't like this style either, but flake8 requires spaces around `=` after a type annotation. If I defy this rule, flake8 raises E252 "missing whitespace around parameter equals." (See https://github.com/PyCQA/pycodestyle/issues/760) It is true this error code is not listed here: https://pycodestyle.pycqa.org/en/latest/intro.html#error-codes but the latest code https://github.com/PyCQA/pycodestyle/blob/master/pycodestyle.py#L1000 actually checks it.